io/src/main/java/io/smallrye/common/io/Files2.java

```diff
@@ -0,0 +1,241 @@
+package io.smallrye.common.io;
+
+import static io.smallrye.common.constraint.Assert.checkNotNullParam;
+import static io.smallrye.common.io.Messages.msg;
+
+import java.io.IOException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.LinkOption;
+import java.nio.file.NotDirectoryException;
+import java.nio.file.Path;
+import java.nio.file.SecureDirectoryStream;
+import java.nio.file.attribute.BasicFileAttributeView;
+
+/**
+ * Extra utilities for dealing with the filesystem which are missing from {@link Files}.
+ */
+public final class Files2 {
+    private Files2() {
+    }
+
+    /**
+     * Open a new secure directory stream at the given path.
+     * If the link option {@link LinkOption#NOFOLLOW_LINKS} is given, the open will fail with an exception
+     * if {@code path} is a symbolic link.
+     *
+     * @param path the path of the directory to open (must not be {@code null})
+     * @param linkOptions the link options (must not be {@code null})
+     *
+     * @return a secure directory stream for the given path (not {@code null})
+     *
+     * @throws NotDirectoryException if the target {@code path} is not a directory
+     * @throws IOException if another I/O error occurs
+     * @throws UnsupportedOperationException if this platform does not support secure directory streams
+     *
+     * @see SecureDirectoryStream#newDirectoryStream(Object, LinkOption...)
+     */
+    public static SecureDirectoryStream<Path> newSecureDirectoryStream(Path path, LinkOption... linkOptions)
+            throws IOException {
+        checkNotNullParam("path", path);
+        checkNotNullParam("linkOptions", linkOptions);
+        path = path.normalize();
+        boolean noFollow = false;
+        for (LinkOption linkOption : linkOptions) {
+            if (linkOption == LinkOption.NOFOLLOW_LINKS) {
+                noFollow = true;
+                break;
+            }
+        }
+        return noFollow ? newSecureDirectoryStreamNoFollow(path) : newSecureDirectoryStreamFollow(path);
+    }
+
+    /**
+     * Attempt to recursively delete the file or directory at the given path,
+     * even if secure directory streams are not supported.
+     * If the target path is a symbolic link, it will be removed.
+     * <em>Warning:</em> this method can potentially delete files outside the intended path
+     * if the target platform does not support secure directory iteration.
+     *
+     * @param path the path to delete (must not be {@code null})
+     * @throws IOException if one or more files fails to be deleted or another I/O error occurs
+     *
+     * @see #deleteRecursively(Path)
+     */
+    public static void deleteRecursivelyEvenIfInsecure(Path path) throws IOException {
+        checkNotNullParam("path", path);
+        path = path.normalize();
+        if (!Files.exists(path)) {
+            return;
+        }
+        Path parent = path.getParent();
```

Comment:
I think a test would be nice, but AFAIU wouldn't this delete the parent of a directory? eg. if you pass `/etc/foo` (`foo` being a directory) it will delete the `/etc` directory recursively?

Reply:
Tests will come next, buts they're a bit more complex for things like this so I want to be sure that we are happy with the api first.

This method opens the directory of the parent and then only recursively deletes the one file inside the parent below using the secure methods. This is because there is no option to not follow symlinks when opening a directory normally.